service/matching/tasklist/adaptive_scaler.go

```diff
@@ -177,91 +192,149 @@ func (a *adaptiveScalerImpl) getPartitionConfig() *types.TaskListPartitionConfig
-			break
+	result := make(map[int]*types.TaskListPartition, targetWritePartitions)
+
+	for i := 0; i < targetWritePartitions; i++ {
+		if p, ok := writePartitions[i]; ok {
+			result[i] = p
+		} else {
+			result[i] = &types.TaskListPartition{}
 		}
-		nw := int32(1)
-		if resp.PartitionConfig != nil {
-			nw = resp.PartitionConfig.NumWritePartitions
+	}
+	return result, true
+}
+
+func (a *adaptiveScalerImpl) adjustReadPartitions(m *aggregatePartitionMetrics, oldReadPartitions map[int]*types.TaskListPartition, newWritePartitions map[int]*types.TaskListPartition) (map[int]*types.TaskListPartition, bool) {
+	result := make(map[int]*types.TaskListPartition, len(newWritePartitions))
+	changed := false
+	for id, p := range oldReadPartitions {
+		result[id] = p
+	}
+	for id, p := range newWritePartitions {
+		if _, ok := result[id]; !ok {
+			changed = true
 		}
-		// in order to drain a partition, 2 conditions need to be met:
-		// 1. the backlog size is 0
-		// 2. no task is being added to the partition, which is guaranteed to be true if the partition knows that the number of write partition is less or equal to its partition ID
-		if resp.TaskListStatus.GetBacklogCountHint() == 0 && nw <= i {
-			// if the partition is drained, we can downscale the number of read partitions
-			numReadPartitions = i
+		result[id] = p
+	}
+
+	for i := len(result) - 1; i >= len(newWritePartitions); i-- {
+		if m.byPartition[i].isDrained() {
 			changed = true
+			delete(result, i)
 		} else {
 			break
 		}
 	}
 	if changed {
-		a.logger.Info("adjust read partitions", tag.NumReadPartitions(numReadPartitions), tag.NumWritePartitions(numWritePartitions))
+		a.logger.Info("adjust read partitions", tag.NumReadPartitions(len(result)), tag.NumWritePartitions(len(newWritePartitions)))
+	}
+	return result, changed
+}
+
+func (a *adaptiveScalerImpl) collectPartitionMetrics(config *types.TaskListPartitionConfig) (*aggregatePartitionMetrics, error) {
+	var mutex sync.Mutex
+	results := make(map[int]*types.DescribeTaskListResponse, len(config.ReadPartitions))
+	g := &errgroup.Group{}
+	for p := range config.ReadPartitions {
+		partitionID := p
+		// Skip the root partition, we can just call the method instead of an rpc
+		if p == 0 {
+			continue
+		}
+		g.Go(func() (e error) {
+			defer func() { log.CapturePanic(recover(), a.logger, &e) }()
+			result, e := a.matchingClient.DescribeTaskList(a.ctx, &types.MatchingDescribeTaskListRequest{
+				DomainUUID: a.taskListID.GetDomainID(),
+				DescRequest: &types.DescribeTaskListRequest{
+					TaskListType: a.taskListType,
+					TaskList: &types.TaskList{
+						Name: a.taskListID.GetPartition(partitionID),
+						Kind: types.TaskListKindNormal.Ptr(),
+					},
+					IncludeTaskListStatus: true,
+				},
+			})
+			if e != nil {
+				a.logger.Warn("failed to get partition metrics", tag.WorkflowTaskListName(a.taskListID.GetPartition(partitionID)), tag.Error(e))
+			}
+			if result != nil {
```

Comment:
if `e` is not nil we shouldn't care about `result`.

Reply:
Done.